### PR TITLE
otelmongo: Use a mock deployment for testing against a MongoDB server

### LINKED ISSUE #39: Add testing to the MongoDB plugin
Create unit testing that doesn't need a running instance of Mongo to function.

_Originally posted by @MrAlias in https://github.com/open-telemetry/opentelemetry-go-contrib/pull/26/review_comment/create_

### PR DESCRIPTION
**Pull request by prestonvasquez**
Resolves #39 

Use mtest to mock a mongodb deployment.

### CODE REVIEW

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go [hunk 1]
```diff
@@ -0,0 +1,234 @@
+// Copied from https://github.com/mongodb/mongo-go-driver/blob/d46f29c34e9ff3968a5c00dcdea8fb41a20a2b2f/mongo/integration/mtest/opmsg_deployment.go
```
Comment (prestonvasquez) [inline instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go:1]: The [precommit target for checking licenses](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/9ddea00fc1eaca6c408db9fa4232831730a6c640/Makefile#L189) doesn't appear to account for the case where code is copied from a third party. Should `license-check` be extended to ignore files with a third party copyright?

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
```
Comment (prestonvasquez) [inline instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go:1]: This file was copied to avoid taking a dependency on the `mtest` package, which is being removed in 2.0.  See [GODRIVER-2724](https://jira.mongodb.org/browse/GODRIVER-2724) fore more information.

File: CHANGELOG.md [hunk 3]
```diff
@@ -39,6 +39,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 
 - Improve performance of `go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc` with the usage of `WithAttributeSet()` instead of `WithAttribute()`. (#5664) 
 - Improve performance of `go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp` with the usage of `WithAttributeSet()` instead of `WithAttribute()`. (#5664) 
+- Copy [opmsg_deployment.go](https://github.com/mongodb/mongo-go-driver/blob/d46f29c34e9ff3968a5c00dcdea8fb41a20a2b2f/mongo/integration/mtest/opmsg_deployment.go) from the mongo-go-driver mtest package to create a mock client for testing otelmongo. Remove all helpers specific to running the `test-mongo-driver` integration test. (#39)
```
Comment (dmathieu) [inline CHANGELOG.md:42]: As this only changes tests and not runtime code/APIs, a changelog entry shouldn't be necessary.

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go (hunk 2 above)
Comment (prestonvasquez) [reply instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go:1]: Here is a blog post describing how to use`mtest` to mock, if the team would rather go this way: https://medium.com/@victor.neuret/mocking-the-official-mongo-golang-driver-5aad5b226a78

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go (hunk 1 above)
Comment (prestonvasquez) [reply instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go:1]: For now I've extended the`license-check` make target and codecov.yaml file to account for this case. See comment concerning mtest

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go [hunk 4]
```diff
@@ -106,11 +102,25 @@ func TestDBCrudOperation(t *testing.T) {
 				otelmongo.WithCommandAttributeDisabled(tc.excludeCommand),
 			)
 			opts.ApplyURI(addr)
+
+			mock := newMockDeployment()
+
+			// nolint:staticcheck
+			//
+			// Deployment is not part of the stable API guaruntee of the
```
Comment (dmathieu) [inline instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go:110]:
```suggestion
			// Deployment is not part of the stable API guarantee of the
```

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go [hunk 5]
```diff
@@ -203,11 +221,25 @@ func TestDBCollectionAttribute(t *testing.T) {
 				otelmongo.WithCommandAttributeDisabled(true),
 			)
 			opts.ApplyURI(addr)
+
+			mock := newMockDeployment()
+
+			// nolint:staticcheck
+			//
+			// Deployment is not part of the stable API guaruntee of the
```
Comment (dmathieu) [inline instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go:229]:
```suggestion
			// Deployment is not part of the stable API guarantee of the
```

**Review by dmathieu — COMMENTED** [review]
This is looking good to me, with the remaining question of the license.
I'll bring this topic to the SIG meeting later today.

**Comment by dmathieu** [comment]
We discussed the licensing issue during the SIG meeting on thursday, and I brought the question to the [#otel-maintainers](https://cloud-native.slack.com/archives/C01NJ7V1KRC/p1718371773724319) slack channel.

What it seems other languages are doing is include both licenses, the original one as well as the OpenTelemetry one.
See [java-instrumentation](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-java-instrumentation+%22Includes+work+from%22&type=code), or [the collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/telemetrygen/main.go).

Would having something similar work for you?

**Comment by prestonvasquez** [comment]
> We discussed the licensing issue during the SIG meeting on thursday, and I brought the question to the [#otel-maintainers](https://cloud-native.slack.com/archives/C01NJ7V1KRC/p1718371773724319) slack channel.
> 
> What it seems other languages are doing is include both licenses, the original one as well as the OpenTelemetry one. See [java-instrumentation](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-java-instrumentation+%22Includes+work+from%22&type=code), or [the collector](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/cmd/telemetrygen/main.go?rgh-link-date=2024-06-17T07%3A40%3A21Z).
> 
> Would having something similar work for you?

@blink1073 @matthewdale Do you have thoughts on this? This PR proposes copying code from [opmsg_deployment.go](https://github.com/mongodb/mongo-go-driver/blob/d46f29c34e9ff3968a5c00dcdea8fb41a20a2b2f/mongo/integration/mtest/opmsg_deployment.go) to mock a server for testing. The ask is that we include both our license as well as the OpenTelemetry license in that copy.

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go (hunk 1 above)
Comment (MrAlias) [reply instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go:1]: The OTel license header needs to added to the top of this file. The existing licences header needs to remain below the added OTel license header and the copied comment can also be added below it.

See [this](https://github.com/open-telemetry/opentelemetry-go/blob/6d45f283c7abf2ee5284c769f9ae7109096d3564/sdk/log/ring.go#L1-L16) for an example.

File: .github/codecov.yaml [hunk 6]
```diff
@@ -15,3 +15,7 @@ comment:
   layout: "reach,diff,flags,tree"
   behavior: default
   require_changes: yes
+
+ignore:
+  # opmsg_deployment is copied from mongo-go-driver.
+  - "instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go"
```
Comment (MrAlias) [inline .github/codecov.yaml:21]: Copying a file from an external project does not exclude it from our test coverage goals. This should not be ignored, and tests should be added to validate the added functionality.

```suggestion
```

**Comment by blink1073** [comment]
>  The ask is that we include both our license as well as the OpenTelemetry license in that copy.

If possible, I'd recommend keeping the original LICENSE and adding a [THIRD-PARTY-NOTICES](https://github.com/mongodb/mongo-go-driver/blob/v1/THIRD-PARTY-NOTICES) file.

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go (hunk 1 above)
Comment (pellared) [reply instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go:1]: More: https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md#code-attribution

Comment (prestonvasquez) [reply instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/opmsg_deployment.go:1]: @MrAlias @pellared @dmathieu Quick update: we are still awaiting feedback concerning the licensing on our end. For simplicity, it may be simpler to just the `mtest` solution. Apologies for the delay on this.

**Comment by jtazin** [comment]
Hi @dmathieu @pellared @MrAlias, @prestonvasquez has reworked the code to import `mtest` in order to satisfy both OpenTelemetry's and MongoDB's licensing requirements. Could you please take a look and let us know if it looks good to you?

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -138,48 +138,3 @@ jobs:
         run: |
           echo ${{ needs.compatibility-test.result }}
           test ${{ needs.compatibility-test.result }} == "success"
-
-  integration:
```
Comment (dmathieu) [inline .github/workflows/ci.yml:142]: We will need to bring this back, as it runs tests for every instrumentation.

**Review by dmathieu — COMMENTED** [review]
This looks nice.

File: .github/workflows/ci.yml (hunk 7 above)
Comment (prestonvasquez) [reply .github/workflows/ci.yml]: I just reverted the change. What do you want to do about the test matrix: https://github.com/open-telemetry/opentelemetry-go-contrib/blob/bb3cf243ee89f724bcab1da580323cbe775505da/.github/workflows/ci.yml#L145

Comment (dmathieu) [reply .github/workflows/ci.yml]: Ah, my bad, you're right. I read this step too quickly, and didn't see it was only for the mongo driver.
Then, yes, it needs to be removed.

**Review by dmathieu — APPROVED** [review]

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go [hunk 8]
```diff
@@ -5,32 +5,28 @@ package test
...
 
 	"github.com/stretchr/testify/assert"
 	"go.mongodb.org/mongo-driver/bson"
 	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/integration/mtest"
 	"go.mongodb.org/mongo-driver/mongo/options"
 
-	"go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo"
-	"go.opentelemetry.io/contrib/internal/util"
+	"go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo" // nolint:staticcheck  // deprecated.
```
Comment (pellared) [inline instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go:17]: Why this change?

```suggestion
	"go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo"
```

File: instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go [hunk 9]
```diff
@@ -106,53 +111,56 @@ func TestDBCrudOperation(t *testing.T) {
...
 			for _, v := range tc.validators {
-				assert.True(t, v(s))
+				assert.True(mt, v(s))
 			}
 		})
 	}
 }
 
 func TestDBCollectionAttribute(t *testing.T) {
+	t.Parallel()
```
Comment (pellared) [inline instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test/mongo_test.go:150]: Can we remove running tests in parallel? 
I do not see any noticeable difference in time execution.
It only increases the complexity of the tests and may make debugging harder.

With `t.Parallel`:
```
$ go test -count=1
PASS
ok      go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test    0.005s
```

Without `t.Parallel`:
```
$ go test -count=1
PASS
ok      go.opentelemetry.io/contrib/instrumentation/go.mongodb.org/mongo-driver/mongo/otelmongo/test    0.005s
```

**Comment by jtazin** [comment]
@pellared could you please take a look at Preston's latest changes? let us know if anything else is needed!

**Review by pellared — APPROVED** [review]
